### PR TITLE
ref(snuba-deletes): introduce delete allocation policies

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context**
Right now we have a simple ratelimiter that limits deletes to `1` concurrent query at a time for the whole endpoint https://github.com/getsentry/snuba/blob/ea2f8e716ea519ae173e7a6833da5d2bf16859f2/snuba/web/views.py#L305

This is good enough for testing out ourselves, but beyond that we wanted to add a more robust ratelimiter that allowed for more granular levers that we have for other endpoints. Product should also be ratelimiting on their end though, this is not supposed to be in lieu of that. 

**delete allocation policies**
We can reuse the allocation policy framework for the delete query pipeline. I've made these policies separate to the other allocation policies definitions because we should be adjusting the limits separately; deletes should not have as high thresholds as select queries because the load those queries cause will be higher, even though they are 'lightweight' (lightweight compared to heavy mutations).

**`DeleteConcurrentRateLimitAllocationPolicy`**
I made a separate class because I want the redis key to be different than the `ConcurrentRateLimitAllocationPolicy` so that we don't clobber ratelimits

This PR just adds the set up for the allocation policies, and does nothing with them (aside from adding they to the `Query` object. The rest of the logic will be in https://github.com/getsentry/snuba/pull/6181

### CODE REVIEW

File: snuba/datasets/configuration/issues/storages/search_issues.yaml [hunk 1]
```diff
@@ -89,6 +89,15 @@ allocation_policies:
         is_enforced: 0
         is_active: 0
 
+delete_allocation_policies:
+  - name: DeleteConcurrentRateLimitAllocationPolicy
+    args:
+      required_tenant_types:
+        - referrer
+        - project_id
```
Comment (armenzg) [inline snuba/datasets/configuration/issues/storages/search_issues.yaml:97]: Would the org be also needed or valuable to have?

**Review by volokluev — COMMENTED** [review]
If you're changing the yaml configuration please add a test for it, otherwise this looks good

**Comment by kylemumma** [comment]
I've never worked with allocation policies before so i'm not super comfortable with this code but as far as I can tell it LGTM.

1 question I have: why did you put `DeleteConcurrentRateLimitAllocationPolicy` under `delete_allocation_policies` rather than under the existing `allocation_policies`

The reason im not approving is bc maybe you should add a test based on Volos comment? Is he referring to the addition of `delete_allocation_policies` in the yaml?

**Review by kylemumma — COMMENTED** [review]
I've never worked with allocation policies before so i'm not super comfortable with this code but as far as I can tell it LGTM.

1 question I have: why did you put `DeleteConcurrentRateLimitAllocationPolicy` under `delete_allocation_policies` rather than under the existing `allocation_policies`

The reason im not approving is bc maybe you should add a test based on Volos comment? Is he referring to the addition of `delete_allocation_policies` in the yaml?

**Comment by MeredithAnya** [comment]
@kylemumma 

> 1 question I have: why did you put `DeleteConcurrentRateLimitAllocationPolicy` under `delete_allocation_policies` rather than under the existing `allocation_policies`

I don't want the delete rate limits to run in the main query pipeline, this is because we will be enforcing different, and more strict limits. I didn't want the logic of getting the allocation policies have to know if the policy is for deletes or not, rather that pipeline is in charge of either getting the select allocation policies or the delete allocation policies. Does that clear things up?

File: snuba/datasets/configuration/issues/storages/search_issues.yaml (hunk 1 above)
Comment (MeredithAnya) [reply snuba/datasets/configuration/issues/storages/search_issues.yaml:97]: yeah I can add it in

**Comment by kylemumma** [comment]
> @kylemumma
> 
> > 1 question I have: why did you put `DeleteConcurrentRateLimitAllocationPolicy` under `delete_allocation_policies` rather than under the existing `allocation_policies`
> 
> I don't want the delete rate limits to run in the main query pipeline, this is because we will be enforcing different, and more strict limits. I didn't want the logic of getting the allocation policies have to know if the policy is for deletes or not, rather that pipeline is in charge of either getting the select allocation policies or the delete allocation policies. Does that clear things up?

@MeredithAnya Yeah that makes sense, thanks for the explanation

**Review by kylemumma — APPROVED** [review]

**Pull request merged**